File: js/core/embedding.ts
export const vec_dim = 64
const token_re = /[^\p{L}\p{N}]+/u

export type vec = number[]

export const seg_text = (txt: string) => {
  const bits = txt
    .split(/[\.\!\?\n\r]+/g)
    .map(v => v.trim().toLowerCase())
    .filter(Boolean)
  if (!bits.length) bits.push(txt.toLowerCase())
  return bits
}

export const embed = (txt: string): vec => {
  const v = new Array(vec_dim).fill(0)
  const toks = txt.toLowerCase().split(token_re).filter(Boolean)
  if (!toks.length) return v
  for (const t of toks) {
    let h = 0
    for (let i = 0; i < t.length; i++) {
      h = (h * 31 + t.charCodeAt(i)) >>> 0
    }
    const idx = h % vec_dim
    v[idx] += 1
    const idx2 = (h >>> 3) % vec_dim
    v[idx2] += 0.5
  }
  const mag = Math.sqrt(v.reduce((s, x) => s + x * x, 0))
  if (mag === 0) return v
  for (let i = 0; i < vec_dim; i++)v[i] /= mag
  return v
}

export const cosine = (a: vec, b: vec) => {
  let dot = 0, ma = 0, mb = 0
  for (let i = 0; i < vec_dim; i++) {
    dot += a[i] * b[i]
    ma += a[i] * a[i]
    mb += b[i] * b[i]
  }
  if (ma === 0 || mb === 0) return 0
  return dot / (Math.sqrt(ma) * Math.sqrt(mb))
}

export const normalize = (v: number) => v < 0 ? 0 : v > 1 ? 1 : v

File: py/core/embedding.py
from __future__ import annotations

import math
import re
from typing import List

VEC_DIM = 64
TOKEN_RE = re.compile(r"[^\w]+", re.UNICODE)


def seg_text(text: str) -> List[str]:
    bits = [s.strip().lower() for s in re.split(r"[.!?\n\r]+", text) if s.strip()]
    return bits or [text.lower()]


def embed(text: str) -> List[float]:
    vec = [0.0] * VEC_DIM
    tokens = [t for t in TOKEN_RE.split(text.lower()) if t]
    if not tokens:
        return vec
    for tok in tokens:
        h = 0
        for ch in tok:
            h = (h * 31 + ord(ch)) & 0xFFFFFFFF
        idx = h % VEC_DIM
        vec[idx] += 1
        idx2 = (h >> 3) % VEC_DIM
        vec[idx2] += 0.5
    mag = math.sqrt(sum(x * x for x in vec))
    if mag:
        vec = [x / mag for x in vec]
    return vec


def cosine(a: List[float], b: List[float]) -> float:
    dot = sum(x * y for x, y in zip(a, b))
    ma = math.sqrt(sum(x * x for x in a))
    mb = math.sqrt(sum(y * y for y in b))
    if ma == 0 or mb == 0:
        return 0.0
    return dot / (ma * mb)


def normalize(value: float) -> float:
    if value < 0:
        return 0.0
    if value > 1:
        return 1.0
    return value
